### ISSUE TITLE
Frame lag between (React) marker and GeoJSON update

### ISSUE THREAD

**Issue by wafisher** [visgl/react-map-gl#940]
(I'm still having issues related to what was supposed to be addressed in #720.)

The code we have will update the position of a train and then propagate that update both directly to GeoJSON and via a `Popup`:

```
<ReactMap
    points={this.state.points}
    onClick={this.onClick}
    onLoad={this.onMapLoad}
    viewport={this.state.viewport}
    updateViewport={this.updateViewport}
>
    <TrainPopup
        train={this.state.selectedTrain}
        close={() => this.setState({ selectedTrain: undefined })}
    />
</ReactMap>
```

The problem is that the React update happens immediately but the GeoJSON update may happen over a couple frames. I've found that large GeoJSON style changes, whether by the old method of setting the style object or the new method of `Source/Layer` will be processed on a worker thread by Mapbox. When it's done (perhaps a few frames later), it will send messages back to the main thread where the change is actually displayed. So the popup will move immediately, but the train icon, which is done in Mapbox via a `Source` doesn't show until later.

Point being here that even calling `render` on Mapbox synchronously may not cause an update in the same frame. The update is actually incredibly small so I'm not sure why Mapbox is taking so long to "think" but regardless, this would happen with a larger update so could still be an issue.

This architecture seems to be confirmed by https://github.com/mapbox/mapbox-gl-native/wiki/Threading-structure but some issues in `mapbox-gl-js` imply this work hasn't been done. Not sure how to solve this from here. As the user, I don't have any insight into when Mapbox will actually render the new updates so it's impossible to time it the same with the popup changes.

![ezgif-5-9f7af3888070](https://user-images.githubusercontent.com/3514570/68777563-d796b900-05ff-11ea-8a1b-54dcc839a1a6.gif)

**Comment by wafisher** [comment]
https://github.com/mapbox/mapbox-gl-js/issues/4904 confirms that `setData` is indeed done on another thread. If I check for the event sent back to the main thread, I can probably time this right. Would be nice for this to be built into `react-map-gl` though.

I'm going to try and write a workaround for now.

**Comment by RobJacobson** [comment]
I'm developing something similar, except that it's for ferry boats in the Seattle area with a popup marker.  I use a separate Marker component for each.  (Note that Marker and Popup both derive from the BaseControl parent component.)  

Since I'm using the same component for both, they move in synchronization with one another.  However, there seem to be performance issues with using markers, as discussed in issues https://github.com/uber/react-map-gl/issues/750 and https://github.com/uber/react-map-gl/issues/909, especially on mobile.

For the sake of asking, have you considered using GeoJSON for both the icon and the popup?  They they should move in sync together and avoid the performance issues mentioned above.

**Comment by wafisher** [comment]
I don't think we can use GeoJSON for the popup, unless I'm missing something? Not sure how you would do that.

**Comment by RobJacobson** [comment]
You're probably correct.  I'm inexperienced with GeoJSON.  It would require some hacking around at a minimum.

The alternative would be to use Marker for the train icon, so that you're essentially using BaseControl for positioning both elements (the icon and the popup) simultaneously.  This is equivalent to how I'm currently doing it, which keeps the elements in sync.  If your typical use-case is just showing a single marker and a single popup at any given time (or a few), this should avoid most of the performance issues.

I'm looking for a more-performant solution using GeoJSON.  I'll update if I find one.

**Comment by RobJacobson** [comment]
I looked at the example code for this map-gl demo:
https://docs.mapbox.com/help/tutorials/building-a-store-locator/#add-custom-markers

This demonstrates how to create a popup from a GeoJSON source using mapbox-gl.  The relevant code is in the "stores.features.foreach(...)" function.  This does the following:
1.  Iterates through each feature.
2.  Creates a mapboxgl.Marker instance for each feature.
3.  Conditionally renders a popup for any particular feature (in createPopUp).

For whatever reason, the mapboxgl.Marker instances do not seem to have the performance issues associated with the markers in react-map-gl.  I need to investigate that issue further.  

However, it's possible to do this -- at least -- using mapbox-gl.  We just need to translate this into react-map-gl code (or switch from react-map-gl to mapbox-gl instead).

**Comment by wafisher** [comment]
Ah yes. I forget why I never went down that route. I haven't looked at the react-map-gl code recently but IIRC, it just renders a normal `<div>` at the pixel point representing the lat/long you specify (i.e. it queries Mapbox for that point on the canvas and puts the div there).

This is definitely an option except that react-map-gl doesn't "natively" support it. Also, how would you specify the contents of the popup? I suppose this could also be done by using the react-map-gl reference to the underlying Mapbox instance, but it seems hacky. Better that react-map-gl perf on popups were better sync'd with Mapbox updates.

**Comment by RobJacobson** [comment]
Agreed.  I'm going to spend some time trying to convert markers and popups from react-map-gl to their mapbox-gl equivalents.  I would be a hack, but I want to compare the performance.  I'll send an update in a few days.

**Comment by RobJacobson** [comment]
@wafisher This is off-topic, but I opened another issue (https://github.com/uber/react-map-gl/issues/977) concerning mobile app development.  Would you mind sharing, in that thread, what framework/tech stack you're using for using react-map-gl in your iOS app?

**Comment by Pessimistress** [comment]
This is being addressed in v7.0. Please follow #1646
